pkg/controller/registry/resolver/step_resolver.go

```diff
@@ -84,14 +86,54 @@ func NewOperatorStepResolver(lister operatorlister.OperatorLister, client versio
 	return stepResolver
 }
 
+// replacementChainFailed returns true if the latest entry in a chain of operators is in the failed state.
+func replacementChainFailed(csv *v1alpha1.ClusterServiceVersion, csvToReplacement map[string]*v1alpha1.ClusterServiceVersion) bool {
+	if csv == nil {
+		return false
+	}
+	if csv.Status.Phase == v1alpha1.CSVPhaseFailed {
+		return true
+	}
+	return replacementChainFailed(csvToReplacement[csv.GetName()], csvToReplacement)
+}
+
 func (r *OperatorStepResolver) ResolveSteps(namespace string) ([]*v1alpha1.Step, []v1alpha1.BundleLookup, []*v1alpha1.Subscription, error) {
 	subs, err := r.listSubscriptions(namespace)
 	if err != nil {
 		return nil, nil, nil, err
 	}
 
+	var omittedCSVs []cache.Predicate
```

Comment:
A little comment explaining what the intent of these predicates are would be helpful when reading through in the future.